libvast/vast/hash_index.hpp

```diff
@@ -255,9 +251,45 @@ class hash_index : public value_index {
   std::vector<digest_type> digests_;
   std::unordered_set<key, key_hasher> unique_digests_;
 
-  // TODO: Once we can use C++20 and have a standard library that implements
-  // key equivalence properly, we can switch to std::unordered_map.
-  detail::steady_map<data, size_t> seeds_;
+  struct data_hash {
+    size_t operator()(const data& x) const {
+      // The default hash computation for `data` and `data_view` is subtly
+      // different: For `data_view` a hash of the contents of the view is
+      // created (so that the same data compares equal whether it's stored
+      // in a default/msgpack/arrow view), but for `data` the hashing is
+      // forwarded to the actual container classes, which can define their own
+      // hash functions.
+      // To ensure the same method is used consistently, we create a view here.
+      return vast::uhash<vast::xxhash>{}(make_view(x));
+    };
+    size_t operator()(const data_view& x) const {
+      return vast::uhash<vast::xxhash>{}(x);
+    };
+  };
+
+  struct data_equal {
+    using is_transparent = void; // Opt-in to heterogenous lookups.
+
+    bool operator()(const data& x, const data_view& y) const {
+      return x == y;
+    }
+
+    bool operator()(const data_view& x, const data& y) const {
+      return x == y;
+    }
+
+    bool operator()(const data& x, const data& y) const {
+      return x == y;
+    }
+
+    bool operator()(const data_view& x, const data_view& y) const {
+      return x == y;
+    }
```

Comment:
```suggestion
    bool operator()(const auto& x, const auto& y) const {
      return x == y;
    }
```

Reply:
Nevermind, that's C++20. You could still use templates to reduce it to just a single function, though.